### ISSUE TITLE
RuntimeTests varied failures

### ISSUE THREAD

**Issue by joestringer** [cilium/cilium#10867]
* Suite-runtime.RuntimeKVStoreTest Consul KVStore
* RuntimePrivilegedUnitTests Run Tests
* RuntimeConntrackInVethModeTest Conntrack-related configuration options for endpoints

https://jenkins.cilium.io/job/Ginkgo-CI-Tests-Pipeline/11057/

https://jenkins.cilium.io/job/Ginkgo-CI-Tests-Pipeline/11067/

https://jenkins.cilium.io/job/Ginkgo-CI-Tests-Pipeline/11071/

https://jenkins.cilium.io/job/Ginkgo-CI-Tests-Pipeline/11050/testReport/junit/(root)/Suite-runtime/RuntimeConntrackInVethModeTest_Conntrack_related_configuration_options_for_endpoints/

**Comment by joestringer** [comment]
I dug into this one:

https://jenkins.cilium.io/job/Ginkgo-CI-Tests-Pipeline/11057/testReport/junit/(root)/Suite-runtime/RuntimeKVStoreTest_Consul_KVStore/
[77a86635_RuntimeKVStoreTest_Consul_KVStore.zip](https://github.com/cilium/cilium/files/4441200/77a86635_RuntimeKVStoreTest_Consul_KVStore.zip)


as I was worried about the following messages:
```
Top 5 errors/warnings:
BPF template object creation failed
Error while rewriting endpoint BPF program
endpoint regeneration failed
Command execution failed
Regeneration of endpoint failed
```

This is somewhat "expected" when Cilium is instructed to shut down while compiling, cf these logs from Cilium:

```
Apr 05 06:29:00 runtime systemd[1]: Stopping cilium...
Apr 05 06:29:00 runtime cilium-agent[4952]: level=error msg="Command execution failed" cmd="[llc -march=bpf -mcpu=v1 -filetype=obj -o /var/run/cilium/state/templates/272162d8b1e4fd4c60ebbc8ecf37d8947d7fa21b/bpf_lxc.o]" error="signal: terminated" subsys=datapath-loader
Apr 05 06:29:00 runtime cilium-agent[4952]: level=error msg="Failed to compile bpf_lxc.o: signal: terminated" compiler-pid=5491 linker-pid=5492 subsys=datapath-loader
Apr 05 06:29:00 runtime cilium-agent[4952]: level=warning msg="JoinEP: Failed to compile" debug=false error="Failed to compile bpf_lxc.o: signal: terminated" params="&{Source:bpf_lxc.c Output:bpf_lxc.o OutputType:obj}" subsys=datapath-loader
Apr 05 06:29:00 runtime cilium-agent[4952]: level=error msg="BPF template object creation failed" bpfHeaderfileHash=272162d8b1e4fd4c60ebbc8ecf37d8947d7fa21b error="failed to compile template program /var/run/cilium/state/templates/272162d8b1e4fd4c60ebbc8ecf37d8947d7fa21b: Failed to compile bpf_lxc.o: signal: terminated" subsys=datapath-loader
Apr 05 06:29:00 runtime cilium-agent[4952]: level=warning msg="BPF template compilation unsuccessful" bpfHeaderfileHash=272162d8b1e4fd4c60ebbc8ecf37d8947d7fa21b error="Could not locate previously compiled BPF template" subsys=datapath-loader
Apr 05 06:29:00 runtime cilium-agent[4952]: level=error msg="Error while rewriting endpoint BPF program" containerID= datapathPolicyRevision=0 desiredPolicyRevision=1 endpointID=343 error="Could not locate previously compiled BPF template" identity=4 ipv4=10.15.156.59 ipv6="f00d::a0f:0:0:1b7d" k8sPodName=/ subsys=endpoint
```

Maybe we can work on improving these messages to take into account the fact that the context canceled and therefore Cilium is shutting down, and that's why the failure occurred.

The next question is why is Cilium shutting down? Not sure if this is part of the test or if there's something else anomalous happening that is triggering this behaviour.

**Comment by pchaigno** [comment]
**TL;DR** I think these flakes were introduced by #10019 and fixed by #10861 (tracked by issue #10857). That bug was merged in master on March 17th and fixed on April 9, which would match the report for the present flakes. I'm closing this; please open a new issue with artifacts if you see these flakes again.

-----

**Details.** #10019 introduced sporadic (underlying concurrency issue) corruptions of `bpf_features.h` resulting in failures to compile the BPF programs. The call to `init.sh` then fails, which, on startup, leads to an abrupt exit of `cilium-agent`.

In details:
`cilium-agent` calls `runDaemon()`:
https://github.com/cilium/cilium/blob/2ff7193bb2301523af39695646ac4eb2a1294f17/daemon/cmd/daemon_main.go#L113
`runDaemon()` calls `NewDaemon()`:
https://github.com/cilium/cilium/blob/2ff7193bb2301523af39695646ac4eb2a1294f17/daemon/cmd/daemon_main.go#L1242
`NewDaemon()` calls `Daemon.init()`:
https://github.com/cilium/cilium/blob/2ff7193bb2301523af39695646ac4eb2a1294f17/daemon/cmd/daemon.go#L503
`Daemon.init()` calls `Loader.Reinitialize()` which itself executes `bpf/init.sh`:
https://github.com/cilium/cilium/blob/2ff7193bb2301523af39695646ac4eb2a1294f17/daemon/cmd/daemon.go#L210

The abrupt exit of `cilium-agent` may lead to the shutdown Joe reported above (`Stopping cilium...`). It also causes other Go routines to stop, such as the compilation of `bpf_lxc.o` seen above.

To validate this, I simulated the behavior of the `bpf_features.h` corruption with a higher probability in #11367. Failing tests and associated logs match the report from Joe.
